File: python/algo/202510/algo3298.py
from collections import Counter, defaultdict

# 滑动窗口与双指针 - 不定长滑动窗口 2.3.2 越长越合法
class Solution:
    def validSubstringCount(self, word1: str, word2: str) -> int:
        # 自行解答
        cnt2 = Counter(word2)
        cnt1 = defaultdict(int)

        # 判断是否包含，性能较差
        def contain() -> bool:
            for k, v in cnt2.items():
                if cnt1[k] < v: return False
            return True

        l = ans = 0
        for ch in word1:
            cnt1[ch] += 1
            cur = True
            while cur and contain():
                cnt1[word1[l]] -= 1
                cur = cnt1[word1[l]] >= cnt2[word1[l]]
                l += 1
            ans += l
        return ans
    
    
    def validSubstringCount2(self, word1: str, word2: str) -> int:
        # 自行解答 - 优化
        cnt2 = Counter(word2)
        cnt1 = defaultdict(int)

        # 有 match2 的长度代替 contain 函数
        l = ans = match2 = 0
        for ch in word1:
            cnt1[ch] += 1
            if cnt1[ch] <= cnt2[ch]: match2 += 1
            while match2 == len(word2):
                cnt1[word1[l]] -= 1
                if cnt1[word1[l]] < cnt2[word1[l]]: match2 -= 1
                l += 1
            ans += l
        return ans

File: cplus/leetcode/202510/algo3298.cpp
#include <bits/stdc++.h>
using namespace std;

// 滑动窗口与双指针 - 不定长滑动窗口 2.3.2 越长越合法
class Solution {
public:
    long long validSubstringCount(string word1, string word2) {
        // translated from python bu doubao 
        unordered_map<char, int> cnt2;
        for (char ch : word2) {
            cnt2[ch]++;
        }
        
        // 统计当前窗口中word1字符的出现次数
        unordered_map<char, int> cnt1;
        int l = 0;
        long long ans = 0;
        
        // 判断当前窗口是否包含word2的所有字符（保持Python原逻辑）
        auto contain = [&]() -> bool {
            for (auto& [k, v] : cnt2) {
                if (cnt1[k] < v) {
                    return false;
                }
            }
            return true;
        };
        
        for (char ch : word1) {
            cnt1[ch]++;
            bool cur = true;
            // 保持原Python的循环条件和逻辑
            while (cur && contain()) {
                char left_ch = word1[l];
                cnt1[left_ch]--;
                // 判断当前左边界字符是否仍满足数量要求
                cur = (cnt1[left_ch] >= cnt2[left_ch]);
                l++;
            }
            ans += l;
        }
        
        return ans;
    }

    long long validSubstringCount2(string word1, string word2) {
        // translated from python bu doubao 
        unordered_map<char, int> cnt2;
        for (char ch : word2) {
            cnt2[ch]++;
        }
        
        unordered_map<char, int> cnt1;
        int l = 0;
        long long ans = 0;
        int match2 = 0;
        int len_word2 = word2.size();
        
        for (char ch : word1) {
            cnt1[ch]++;
            // 只有当当前计数仍小于等于目标计数时，才增加匹配数
            if (cnt1[ch] <= cnt2[ch]) {
                match2++;
            }
            
            // 当匹配数达到word2长度时，尝试收缩左边界
            while (match2 == len_word2) {
                char left_ch = word1[l];
                cnt1[left_ch]--;
                // 如果收缩后计数小于目标，减少匹配数
                if (cnt1[left_ch] < cnt2[left_ch]) {
                    match2--;
                }
                l++;
            }
            
            // 累加有效子字符串数量
            ans += l;
        }
        
        return ans;
    }

};